Vendor cluster capacity with origin build tooling

aveshagarwal opened the pull request:
It does:
Builds cluster capacity with origin tooling.
Updates origin.spec for building cluster capacity as a sub rpm.
Creates cluster capacity image from the sub rpm.

I am planning to do some clean up, but would be good to have some feedback. 

@derekwaynecarr @stevekuznetsov @eparis @jpeeler 
Deprecates the previous PR: https://github.com/openshift/origin/pull/14014

stevekuznetsov commented:
Last commit is blocked on #13642 -- we will need to do the build-time mount for that. For now, use ADD to get them into the image. I would also have expected that image manifest to live in the fork.

aveshagarwal commented:
@stevekuznetsov @derekwaynecarr @jpeeler

So here is the current update:
1. The cluster capacity vendoring in cmd/cluster-capacity/src/github.com/kubernetes-incubator/cluster-capacity/  is based on https://github.com/openshift/cluster-capacity/pull/3 .
2. The current directory cmd/cluster-capacity/src/github.com/kubernetes-incubator/cluster-capacity/ has build-go.sh and build-cross.sh are able to create binaries.
3. The current directory cmd/cluster-capacity/src/github.com/kubernetes-incubator/cluster-capacity/ also has build-images.sh and dockefile in images/cluster-capacity that creates cluster capacity image.

So what else is needed?

stevekuznetsov commented:
Depends on what you want the image build in the fork to do without Origin -- if you want to be able to build _an_ image, continue to use the current approach. If you want it to only work with the RPMs, change it to be `FROM openshift/origin-source` and make the binary install use `yum` like the [`origin-pod` Dockerfile](https://github.com/openshift/origin/blob/master/images/pod/Dockerfile#L9-L14):

```Dockerfile
FROM openshift/origin-source

RUN INSTALL_PKGS="origin-pod" && \
    yum install -y ${INSTALL_PKGS} && \
    rpm -V ${INSTALL_PKGS} && \
    yum clean all
```

In any case, the Dockerfile used to build the image in Origin will need that flow. In this PR you'll probably want to add an environment variable like `$OS_BUILD_INCUBATOR_IMAGES=true` to `hack/build-images.sh` or make a `hack/build-incubator-images.sh` script to build your images once the `origin` RPM package is built. I'd make the default behavior of `make release` build these images.

aveshagarwal commented:
@sdodson

aveshagarwal commented:
@stevekuznetsov I have added Dockerfile in origin/images/cluster-capacity like origin/images/pod to build cluster capacity image from origin-cluster-capacity rpm. PTAL.

derekwaynecarr commented:
@aveshagarwal -- is this good to go?

derekwaynecarr commented:
[test]

aveshagarwal commented:
@derekwaynecarr so i git archived and then extracted the tar at the current location. so its an exact copy of openshift/cluster-capacity but without .git dir stuff.

derekwaynecarr commented:
@aveshagarwal -- checkout the test failures.

aveshagarwal commented:
Seems like go vet is failing on vendored stuff in cluster capacity, some part of the logs:

```
cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/gopkg.in/yaml.v2/encode_test.go:344: struct field tag ",inline" not compatible with reflect.StructTag.Get: bad syntax for struct tag pair
cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/gopkg.in/yaml.v2/encode_test.go:350: struct field tag ",inline" not compatible with reflect.StructTag.Get: bad syntax for struct tag pair
cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/gopkg.in/yaml.v2/encode_test.go:406: struct field tag "_" not compatible with reflect.StructTag.Get: bad syntax for struct tag pair
cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/k8s.io/apimachinery/pkg/runtime/scheme_test.go:586: no formatting directive in Errorf call
cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/scheme_test.go:586: no formatting directive in Errorf call
FAILURE: go vet failed!
````

Will look into it soon.

aveshagarwal commented:
Its weird that one of the errors is following, when in fact, cmd/hypercc does have buildable go source files.

```
++ Building go targets for linux/amd64: ../../cmd/hypercc
can't load package: package github.com/kubernetes-incubator/cluster-capacity/cmd/hypercc: no buildable Go source files in /openshifttmp/openshift/build-rpm-release/tito/rpmbuild-origin87gSss/BUILD/origin-3.6.0/cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/_output/local/go/src/github.com/kubernetes-incubator/cluster-capacity/cmd/hypercc
```

stevekuznetsov commented:
@aveshagarwal you'll want to black-list the entire `$GOPATH` for cluster-capacity [here](https://github.com/openshift/origin/blob/master/hack/verify-govet.sh#L25).

aveshagarwal commented:
@stevekuznetsov thanks, I will do this. Do you know why this error:

```
++ Building go targets for linux/amd64: ../../cmd/hypercc
can't load package: package github.com/kubernetes-incubator/cluster-capacity/cmd/hypercc: no buildable Go source files in /openshifttmp/openshift/build-rpm-release/tito/rpmbuild-origin87gSss/BUILD/origin-3.6.0/cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/_output/local/go/src/github.com/kubernetes-incubator/cluster-capacity/cmd/hypercc
```

When I build rpm locally from origin.spec, it works fine. The above error seems weird as cmd/hypercc does have buildable go source files.

stevekuznetsov commented:
@aveshagarwal can you do a full log of whatever is going on with `OS_DEBUG=true`? I don't have enough context to say how your set up locally is different than the one that is failing.

aveshagarwal commented:
I have addressed 2 issues: rpm building and gofmt, still looking into go vet.

aveshagarwal commented:
@stevekuznetsov the last commit changes address removing cmd/cluster-capacity from go vet list. PTAL.

aveshagarwal commented:
[test]

stevekuznetsov commented:
@aveshagarwal I would suggest instead to edit the source for `find_files` [here](https://github.com/openshift/origin/blob/master/hack/util.sh#L19) to add an exception for the base of your `$GOPATH`.

aveshagarwal commented:
@stevekuznetsov I am not sure how would that work unless I am missing something. As we want to only partially cut cmd/ (by just removing cmd/cluster-capacity not other files/dirs in cmd/) not completely, so find_files would always end up cmd in verify-govet.sh and that mean it would keep vetting cmd/cluster-capacity.

stevekuznetsov commented:
@aveshagarwal The base of your `$GOPATH` is `cmd/cluster-capacity/go` -- not sure what you mean

@derekwaynecarr @smarterclayton @pmorie why did we put this whole tree under `cmd/`?

aveshagarwal commented:
@stevekuznetsov hack/verify-govet.sh has following:

test_dirs="$(find_files | cut -d '/' -f 1-2 | sort -u)"

Even if we remove cmd/cluster-capacity/go from find_files, the above test_dirs would have cmd that means would keep vetting cmd/cluster-capacity.

stevekuznetsov commented:
@aveshagarwal this diff would suffice

```patch
From 48c477300517014641396fa06fd9be5e1ed4c811 Mon Sep 17 00:00:00 2001
From: Steve Kuznetsov <skuznets@redhat.com>
Date: Wed, 24 May 2017 13:03:57 -0700
Subject: [PATCH] Exclude cluster-capacity source files from source list

We do not want the cluster-capacity source files to be searched by all
of the scripts that currently rely on the `find_files` utility, like the
scripts that run `go fmt`, `go vet` and `go lint.`

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>
---
 hack/util.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/hack/util.sh b/hack/util.sh
index 0286f4c..9d15ce4 100644
--- a/hack/util.sh
+++ b/hack/util.sh
@@ -16,6 +16,7 @@ function find_files() {
 		-o -wholename './pkg/bootstrap/bindata.go' \
 		-o -wholename './openshift.local.*' \
 		-o -wholename './test/extended/testdata/bindata.go' \
+		-o -wholename './cmd/cluster-capacity/*' \
 		-o -wholename '*/vendor/*' \
 		-o -wholename './assets/bower_components/*' \
 		\) -prune \
-- 
2.9.3

```

stevekuznetsov commented:
Ah, I see

aveshagarwal commented:
@stevekuznetsov gofmt an golint seem fine with cluster-capacity. So are you ok with current changes for govet?

stevekuznetsov commented:
Seems weird to me that this is under `cmd/` but if that must be the final place for the vendor we will need to do a larger diff. I don't want to regress on the files we run `go vet` on -- removing `cmd/` will remove files other than the ones you vendor in.

stevekuznetsov commented:
As other scripts operate per-file on the output of `find_files` I would _still_ suggest the patch above -- it is necessary but not sufficient for this.

stevekuznetsov commented:
@aveshagarwal you could try adding a regex to the blacklist at the top:

```sh
govet_blacklist=(
    # ...
    "cmd/cluster-capacity/.*"
)
```

aveshagarwal commented:
@stevekuznetsov thanks, it seems that regex is working, so I have update the last commit. PTAL.

aveshagarwal commented:
[test]

derekwaynecarr commented:
[test]

derekwaynecarr commented on origin.spec:
```diff
@@ -256,6 +287,11 @@ install -p -m 755 _output/local/bin/darwin/amd64/oc %{buildroot}/%{_datadir}/%{n
+# Install cluster capacity
+install -p -m 755 cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/_output/local/bin/${PLATFORM}/hypercc %{buildroot}%{_bindir}/
+ln -s hypercc %{buildroot}%{_bindir}/cluster-capacity
+ln -s hypercc %{buildroot}%{_bindir}/genpod
```
is this needed?  its a test tool...  i dont want to support it.

derekwaynecarr replied:
not going to block on this now

derekwaynecarr commented on origin.spec:
```diff
@@ -583,6 +619,12 @@ fi
+%files cluster-capacity
+%{_bindir}/hypercc
+%{_bindir}/cluster-capacity
+%{_bindir}/genpod
```
i dont want to support this binary.

aveshagarwal replied:
i can remove genpod.

derekwaynecarr replied:
not going to block on this now.

aveshagarwal commented:
I doubt that the last test results are after the latest changes. because it is not possible to complete all tests again after the latest changes.

derekwaynecarr commented:
@aveshagarwal -- you can see the commit id in the failure message.  it was for your prior commit.

derekwaynecarr commented:
[test]

aveshagarwal commented:
wonder why the following is failing for `go vet` in this PR, I dont see this failure locally:
```
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/crypto/sha3/keccakf_amd64.s:324: [amd64] keccakF1600: unknown variable state; offset 0 is a+0(FP)
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/sys/plan9/asm.s: cannot determine architecture for assembly file
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/sys/unix/asm.s: cannot determine architecture for assembly file
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/sys/unix/asm_linux_amd64.s:16: [amd64] Syscall: function Syscall missing Go declaration
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/sys/unix/asm_linux_amd64.s:19: [amd64] Syscall6: function Syscall6 missing Go declaration
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/sys/unix/asm_linux_amd64.s:22: [amd64] RawSyscall: function RawSyscall missing Go declaration
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/sys/unix/asm_linux_amd64.s:25: [amd64] RawSyscall6: function RawSyscall6 missing Go declaration
: cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/golang.org/x/sys/unix/asm_linux_amd64.s:28: [amd64] gettimeofday: function gettimeofday missing Go declaration
FAILURE: go vet failed!
```

aveshagarwal commented:
[test]

aveshagarwal commented:
[test]

aveshagarwal commented:
Failure in one test does not seem related to cluster capacity:

```
Ginkgo ran 1 suite in 14m1.198218212s
Test Suite Passed
[INFO] [CLEANUP] Beginning cleanup routines...
[INFO] [CLEANUP] Dumping cluster events to _output/scripts/conformance/artifacts/events.txt
[INFO] [CLEANUP] Dumping container logs to _output/scripts/conformance/logs/containers
chroot: cannot change root directory to /rootfs: No such file or directory
chroot: cannot change root directory to /rootfs: No such file or directory
[INFO] [CLEANUP] Truncating log files over 100M
[INFO] [CLEANUP] Stopping docker containers
chroot: cannot change root directory to /rootfs: No such file or directory
chroot: cannot change root directory to /rootfs: No such file or directory
[INFO] [CLEANUP] Removing docker containers
chroot: cannot change root directory to /rootfs: No such file or directory
chroot: cannot change root directory to /rootfs: No such file or directory
[INFO] [CLEANUP] Killing child processes
[INFO] [CLEANUP] Pruning etcd data directory
rm: cannot remove ‘/tmp/etcd/openshift-backup-etcd_backup_tag20170525015920’: Permission denied
rm: cannot remove ‘/tmp/etcd/openshift-backup-etcd_backup_tag20170525015845’: Permission denied
rm: cannot remove ‘/tmp/etcd/member’: Permission denied
[ERROR] test/extended/conformance.sh exited with code 1 after 00h 47m 13s
make: *** [test-extended] Error 1
++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace

```

aveshagarwal commented:
same with error in another tests:
```
Error response from daemon: Untar error on re-exec cmd: fork/exec /proc/self/exe: cannot allocate memory
```

Seems like last errors were due to flakiness of infra.

aveshagarwal commented:
[test]

aveshagarwal commented:
[test] #13980

aveshagarwal commented:
@derekwaynecarr I am tagging it merge as all tests passed. Though not sure if I have merge rights.

aveshagarwal commented:
[merge]

derekwaynecarr commented:
@aveshagarwal -- we are past sprint dev cut.  it will not merge until master reopens at the start of next sprint.  any decision to merge prior to that point will be based on discussion with @eparis and understanding if the new rpms present risk to other downstream tooling.  that said, thank you and @stevekuznetsov for getting this to pass all tests!

aveshagarwal commented:
@derekwaynecarr @eparis I have updated following things:

1. Rebased to latest origin
2. Removed genpod from origin packaging
3. Updates based on eparis' feedback
4. Updated images/build-image.sh so that cluster capacity is build with origin build flow.

Let me know If I missed something.

eparis commented:
does not build:https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_integration/2698/console

aveshagarwal commented:
ok looking why it failed.

aveshagarwal commented:
I have not changed anything, in origin.spec or build-cross.sh and previously everything passed including build succeeded. 

```
+ unset GOPATH
+ cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/hack/build-cross.sh
++ Building go targets for linux/amd64: cmd/hypercc
pkg/framework/report.go:26:2: cannot find package "github.com/ghodss/yaml" in any of:
	/openshifttmp/openshift/build-rpm-release/tito/rpmbuild-origindrSQE2/BUILD/origin-3.6.0/cmd/cluster-capacity/go/src/github.com/kubernetes-incubator/cluster-capacity/_output/local/go/src/github.com/kubernetes-incubator/cluster-capacity/vendor/github.com/ghodss/yaml (vendor tree)
	/usr/local/go/src/github.com/ghodss/yaml (from $GOROOT)
```
@stevekuznetsov  Any thoughts?

aveshagarwal commented:
seems like the error from (https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin/1854/) which is previous run and after that I have already updated, lets see how the current run 1857 goes.

@stevekuznetsov Is there any way to stop Origin CI to stop running previous run if there is already an update to a PR. It seems to be wasting resources it seems.

stevekuznetsov commented:
> Is there any way for Origin CI to terminate running previous run if there is already an update to a PR. It seems to be wasting resources it seems.

There's probably nothing architectural stopping someone from implementing that feature, but it's just not implemented.

aveshagarwal commented:
@eparis @derekwaynecarr all tests passed.

eparis commented:
[merge] This is going to fail, but will get us at the top of the queue come saturday.

eparis commented:
[merge]

stevekuznetsov commented:
@openshift-bot woooo

smarterclayton commented:
Follow up item: the cluster capacity image is 773M of unshared layers with the rest of OpenShift (i.e. if you download origin onto a node, to get federation you have to get 750M+ of additional content).  That sucks and we need to fix it in a follow up.  At minimum we need to know where that size is coming from and what is common with openshift/origin.  Also, it doesn't inherit from origin-base which is our standard tooling layer.

aveshagarwal commented:
@smarterclayton cluster-capacity's binary is around 67M. So mostly this size is from the base image it is using which is openshift/origin-source and any other rpms it is pulling during image build process.
I was checking build logs and I saw It was pulling several rpm. I will look more into it later. 

```
FROM openshift/origin-source

RUN INSTALL_PKGS="origin-cluster-capacity" && \
    yum install -y ${INSTALL_PKGS} && \
    rpm -V ${INSTALL_PKGS} && \
    yum clean all

LABEL io.k8s.display-name="OpenShift Origin Cluster Capacity" \
      io.k8s.description="This is a component of OpenShift Origin and runs cluster capacity analysis tool."

CMD ["/usr/bin/cluster-capacity --help"]
```

enj commented:
Maybe I missed something, but why did we have to vendor cluster capacity in this strange way?

stevekuznetsov commented:
We needed to ship the `cluster-capacity` binaries as part of the `atomic-openshift` RPM, which means it's got to be part of the `dist-git` repo that creates it, and the simplest way of doing that was having it in this repo. Images in OSBS get their binaries from a `yum install` so we need that intermediary RPM, can't just ship images but not binaries. If we split `cluster-capacity` _et al_ out of our RPM they can stand alone entirely.
